packages/auth-nextjs/src/app/index.ts

```diff
@@ -9,26 +9,53 @@ import { cookies } from "next/headers";
 import { redirect } from "next/navigation";
 import type { NextRequest } from "next/server";
 
-import { NextAuth, NextAuthSession, type NextAuthOptions } from "../shared";
+import {
+  NextAuth,
+  NextAuthSession,
+  type NextAuthOptions,
+  BuiltinProviderNames,
+} from "../shared";
 
-export { NextAuthSession, type NextAuthOptions };
+export {
+  NextAuthSession,
+  type NextAuthOptions,
+  type BuiltinProviderNames,
+  type TokenData,
+};
 
-type ParamsOrError<Result extends object> =
-  | ({ error: null } & Result)
-  | ({ error: Error } & { [Key in keyof Result]?: undefined });
+type ParamsOrError<Result extends object, ErrorDetails extends object = {}> =
+  | ({ error: null } & { [Key in keyof ErrorDetails]?: undefined } & Result)
+  | ({ error: Error } & ErrorDetails & { [Key in keyof Result]?: undefined });
```

Comment:
Not a blocker: does moving the payload into a separate key make it easier for this to be a simple discriminated union?

Something like:
```suggestion
type ParamsOrError<Result extends object, ErrorDetails extends object = {}> =
  | { error: null; params: Result }
  | { error: Error; details: ErrorDetails };
```